testsuite/classlibrary/TestNodeProxy_Server.sc

```diff
@@ -53,6 +53,19 @@ TestNodeProxy_Server : UnitTest {
 		this.assertEquals(build, false, "SynthDef should not be rebuilt when calling send");
 	}
 
+	test_synthDef_isReleased_afterFree {
+		var numBefore, numAfter;
+		numBefore = server.statusWatcher.numSynthDefs;
+		proxy.source = { Silent.ar };
+		server.sync;
+		proxy.source = nil;
+		server.sync;
```

Comment:
Does this test not rely on the status watcher having received a status update from after the def has been freed, but before the sync reply comes? Not sure how likely that is to fail, but seems a potential risk. (Also noticed that /sync is not specifically documented as asynchronous, but perhaps obvious).

Reply:
IIUC the server processes messages in a FIFO loop. since these replies are sent on the same machine and not over a network with unpredictable routing behavior, you're guaranteed that the ordering of messages going out of sclang matches that going into scsynth, matches that going out of scsynth, matches that going into sclang.

Reply:
Yes that's my understanding too, but I'm still worried. If there was an explicit status request sent before the last sync I think it would be fine. To be honest, looking at it again I'm not sure why this test should work. But it's late and I'm tired, and maybe I'm misunderstanding something about the way the language yields, the ServerWatcher status request scheduling and the synchronous d_free interact?

Reply:
maybe you could briefly sketch out the ordering of events as you understand them now and why that doesn't work? i'm having trouble understanding your concern

Reply:
I think Scott's concern is legit.

https://github.com/supercollider/supercollider/blob/develop/server/scsynth/SC_SequencedCommand.cpp#L1104-L1143

/status is not documented as asynchronous, but it's defined as a "sequence command," so it *is* asynchronous and it will cause a /synced to reply to wait. And the /status.reply must be sent before the status command is complete, so I would expect the order of messages scsynth -> sclang to be `/status.reply`, then `/synced`.

*But* this assumes that a /status message is pending in the queue at the time /sync is sent.

Scott is correct that there is nothing in the code here to guarantee this is the case. Actually, since status messages are sent only every 700 ms, it's extremely unlikely that this is the case.

So, suppose something breaks and these synthdefs aren't being deleted. This test will probably not wait for the /status.reply with the updated number of synthdefs -- therefore, it's likely to pass even though the evidence of failure will come a few hundred milliseconds later.

I believe `server.sendStatusMsg.sync` would fix it. If you want /sync to wait for a status message, you're responsible for forcing a status request to be sent before syncing.

Reply:
thanks! that makes sense to me

Reply:
I'm surprised that it worked.

Reply:
Yes, I can't quite figure out why it would have worked. But I am a bear of little brain (or at least little remaining at the end of 2019, LOL). The other piece of this puzzle is where the yields occur that would allow the message to be sent, and how that interacts with the required ordering. Perhaps moot now, but interesting...

Reply:
"I'm surprised that it worked." ... meaning, the test failed without the code fix? I see how the test would pass, don't see how it would fail.